sys/include/net/nanocoap/cache.h

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (C) 2020 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    net_nanocoap_cache Nanocoap-Cache implementation
+ * @ingroup     net_nanocoap
+ * @brief       A cache implementation for nanocoap response messages
+ *
+ * @{
+ *
+ * @file
+ * @brief       nanocoap-cache API
+ *
+ * @author      Cenk Gündoğan <cenk.guendogan@haw-hamburg.de>
+ */
+
+#ifndef NET_NANOCOAP_CACHE_H
+#define NET_NANOCOAP_CACHE_H
+
+#include <assert.h>
+#include <stdint.h>
+#include "clist.h"
+#include "net/nanocoap.h"
+#include "hashes/sha256.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief The number of maximum cache entries.
+ */
+#ifndef CONFIG_NANOCOAP_CACHE_ENTRIES
+#define CONFIG_NANOCOAP_CACHE_ENTRIES          (8)
+#endif
+
+/**
+ * @brief The length of the cache key in bytes.
+ */
+#ifndef CONFIG_NANOCOAP_CACHE_KEY_LENGTH
+#define CONFIG_NANOCOAP_CACHE_KEY_LENGTH       (8)
+#endif
+
+/**
+ * @brief Size of the buffer to store responses in the cache.
+ */
+#ifndef CONFIG_NANOCOAP_CACHE_RESPONSE_SIZE
+#define CONFIG_NANOCOAP_CACHE_RESPONSE_SIZE    (128)
+#endif
+
+/**
+ * @brief   Cache container that holds a @p coap_pkt_t struct.
+ */
+typedef struct {
+    /**
+     * @brief needed for clist_t, must be the first struct member!
+     */
+    clist_node_t node;
+
+    /**
+     * @brief the calculated cache key, see nanocoap_cache_key_generate().
+     */
+    uint8_t cache_key[CONFIG_NANOCOAP_CACHE_KEY_LENGTH];
+
+    /**
+     * @brief packet representation of the response
+     */
+    coap_pkt_t response_pkt;
+
+    /**
+     * @brief buffer to hold the response message.
+     */
+    uint8_t response_buf[CONFIG_NANOCOAP_CACHE_RESPONSE_SIZE];
+
+    size_t response_len; /**< length of the message in @p response */
+
+    unsigned request_method; /**< the method of the initial request */
+
+    /**
+     * @brief absolute system time (us) that records last access to
+     * this cache entry (insertion and lookup).
+     */
+    uint64_t access_time;
+
+    /**
+     * @brief absolute system time (us) until which this cache entry
```

Comment:
Why such a high precision?

Reply:
Fixed with cgundogan#29 merged ;-)

Reply:
Ah, but the doc needs to be adapted. @cgundogan could you please?